src/foam/dao/ArrayDAO.js

```diff
@@ -33,99 +33,158 @@ foam.CLASS({
+        var detached = false;
+        var sub = foam.core.FObject.create();
+        sub.onDetach(function() { detached = true; });
 
-      var self = this;
+        var self = this;
 
-      return new Promise(function(resolve, reject) {
-        for ( var i = 0 ; i < self.array.length ; i++ ) {
-          if ( detached ) break;
+        return new Promise(function(resolve, reject) {
+          for ( var i = 0 ; i < self.array.length ; i++ ) {
+            if ( detached ) break;
 
-          sink.put(self.array[i], sub);
-        }
+            sink.put(self.array[i], sub);
+          }
 
-        sink.eof();
+          sink.eof();
 
-        resolve(resultSink);
-      });
+          resolve(resultSink);
+        });
+      },
+      javaCode: `
+      Sink resultSink = prepareSink(sink);
+      sink = decorateSink(x, resultSink, skip, limit, order, predicate);
+      for ( Object o : getArray() ) {
+        sink.put(o, null);
+      }
+      return resultSink;
+      `
     },
-
-    function removeAll_(x, skip, limit, order, predicate) {
-      predicate = predicate || this.True.create();
-      skip = skip || 0;
-      limit = foam.Number.isInstance(limit) ? limit : Number.MAX_VALUE;
-
-      for ( var i = 0 ; i < this.array.length && limit > 0 ; i++ ) {
-        if ( predicate.f(this.array[i]) ) {
-          if ( skip > 0 ) {
-            skip--;
-            continue;
+    {
+      name: 'removeAll_',
+      code: function(x, skip, limit, order, predicate) {
+        predicate = predicate || this.True.create();
+        skip = skip || 0;
+        limit = foam.Number.isInstance(limit) ? limit : Number.MAX_VALUE;
+
+        for ( var i = 0 ; i < this.array.length && limit > 0 ; i++ ) {
+          if ( predicate.f(this.array[i]) ) {
+            if ( skip > 0 ) {
+              skip--;
+              continue;
+            }
+            var obj = this.array.splice(i, 1)[0];
+            i--;
+            limit--;
+            this.on.remove.pub(obj);
           }
-          var obj = this.array.splice(i, 1)[0];
-          i--;
-          limit--;
-          this.on.remove.pub(obj);
         }
-      }
 
-      return Promise.resolve();
+        return Promise.resolve();
+      }
     },
-
-    function find_(x, key) {
-      var id = this.of.isInstance(key) ? key.id : key;
-      for ( var i = 0 ; i < this.array.length ; i++ ) {
-        if ( foam.util.equals(id, this.array[i].id) ) {
-          return Promise.resolve(this.array[i]);
+    {
+      name: 'find_',
+      code: function(x, key) {
+        // Predicate handled in AbstractDAO
+        if ( this.of.isInstance(key) ) {
+          var expr = this.identityExpr
+          for ( var i = 0 ; i < this.array.length ; i++ ) {
+            if ( foam.util.equals(id.f(this.array[i])) ) {
+              return Promise.resolve(this.array[i]);
+            }
+          }
+        } else {
+          var id = this.of.isInstance(key) ? key.id : key;
```

Comment:
Don't we already know that the this.of.isInstance(key) will return false here?

Reply:
from the console, using a User object, isInstance behaves as expected.